hub/integrations/pytorch/dataset.py

```diff
@@ -0,0 +1,497 @@
+from typing import Callable, Iterable, Optional, Sequence, List, Union
+from hub.constants import MB
+
+from hub.util.iterable_ordered_dict import IterableOrderedDict
+from hub.core.io import (
+    SampleStreaming,
+    Schedule,
+    SequentialMultithreadScheduler,
+    ShufflingSchedulerWrapper,
+    SingleThreadScheduler,
+    MultiThreadedNaiveScheduler,
+)
+from hub.integrations.pytorch.shuffle_buffer import ShuffleBuffer
+
+import torch
+import torch.utils.data
+
+from torch.multiprocessing import Queue, Process
+from torch._utils import ExceptionWrapper
+from torch.utils.data.dataloader import DataLoader
+from torch.utils.data._utils.worker import ManagerWatchdog
+from torch._C import _remove_worker_pids, _set_worker_pids, _set_worker_signal_handlers
+from torch.utils.data._utils.signal_handling import _set_SIGCHLD_handler
+from warnings import warn
+from queue import Empty
+
+import numpy as np
+
+mp = torch.multiprocessing.get_context()
+
+
+def use_scheduler(num_workers: int, ensure_order: bool):
+    if num_workers <= 1:
+        return SingleThreadScheduler()
+    else:
+        if ensure_order:
+            return MultiThreadedNaiveScheduler(num_workers)
+        else:
+            return SequentialMultithreadScheduler(num_workers)
+
+
+def cast_type(tensor: np.ndarray):
+    if tensor.dtype == np.uint16:
+        return tensor.astype(np.int32)
+    if tensor.dtype == np.uint32:
+        return tensor.astype(np.int64)
+    if tensor.dtype == np.uint64:
+        return tensor.astype(np.int64)
+
+    return tensor
+
+
+def _process(tensor, transform: Optional[Callable]):
+    tensor = IterableOrderedDict((k, cast_type(tensor[k])) for k in tensor)
+
+    if transform:
+        tensor = transform(tensor)
+
+    return tensor
+
+
+class _ContinueIteration:
+
+    """
+    Instructs worker to load N samples of data and publish it to the
+    data_queue
+
+    Args:
+        n: int number of samples to publish. Acts as in fly buffer_size similar
+           to reactive stream approach
+    """
+
+    def __init__(self, n: int) -> None:
+        self.n = n
+
+
+class _StopIteration:
+
+    """
+    Indicates worker done with its schedule
+    """
+
+    def __init__(self) -> None:
+        pass
+
+
+def _worker_loop(
+    wid: int,
+    dataset,
+    tensors,
+    use_local_cache: bool,
+    schedule: Schedule,
+    transform: Callable,
+    request_queue: Queue,
+    data_queue: Queue,
+    workers_done,
+):
+    torch.set_num_threads(1)
+    _set_worker_signal_handlers()
+
+    try:
```

Comment:
huge method, difficult to follow, can we break this up?

Reply:
Not really.. it is huge worker loop, everything is interconnected there. It is pretty much the same in `pytorch` implementation.